volatility3/framework/plugins/linux/netfilter.py

```diff
@@ -0,0 +1,714 @@
+    """In 4.9 it was changed to a simple singly-linked list.
+    struct nf_hook_entry * nf_hooks_ingress;
+    """
+
+    @classmethod
+    def symtab_checks(cls, vmlinux) -> bool:
+        hooks_type = ["pointer", "nf_hook_entry"]
+        return (
+            vmlinux.has_symbol("net_namespace_list")
+            and vmlinux.has_type("net_device")
+            and vmlinux.get_type("net_device").has_member("nf_hooks_ingress")
+            and cls.get_member_type(vmlinux.get_type("net_device"), "nf_hooks_ingress")
+            == hooks_type
+        )
+
+    def get_hook_ops(self, hook_container, proto_idx, hook_idx):
+        nf_hooks_ingress_ptr = hook_container
+        if not nf_hooks_ingress_ptr:
+            return
+
+        while nf_hooks_ingress_ptr:
+            nf_hook_entry = nf_hooks_ingress_ptr.dereference()
+            orig_ops = nf_hook_entry.orig_ops.dereference()
+            yield orig_ops
+            nf_hooks_ingress_ptr = nf_hooks_ingress_ptr.next
+
+
+class NetfilterNetDevImp_4_14_to_latest(AbstractNetfilterNetDev):
+    """In 4.14 the hook list was converted to an array of pointers inside the struct
+    'nf_hook_entries':
+    struct nf_hook_entries * nf_hooks_ingress;
+    struct nf_hook_entries {
+            u16 num_hook_entries;
+            struct nf_hook_entry        hooks[];
+            //const struct nf_hook_ops *orig_ops[]; }
+    """
+
+    @classmethod
+    def symtab_checks(cls, vmlinux) -> bool:
+        hooks_type = ["pointer", "nf_hook_entries"]
+        return (
+            vmlinux.has_symbol("net_namespace_list")
+            and vmlinux.has_type("net_device")
+            and vmlinux.get_type("net_device").has_member("nf_hooks_ingress")
+            and cls.get_member_type(vmlinux.get_type("net_device"), "nf_hooks_ingress")
+            == hooks_type
+        )
+
+    def get_hook_ops(self, hook_container, proto_idx, hook_idx):
+        nf_hook_entries = hook_container
+        if not nf_hook_entries:
+            return
+
+        nf_hook_ops_name = self.get_symbol_fullname("nf_hook_ops")
+        nf_hook_ops_ptr_arr = self.build_nf_hook_ops_array(nf_hook_entries)
+        for nf_hook_ops_ptr in nf_hook_ops_ptr_arr:
+            nf_hook_ops = nf_hook_ops_ptr.dereference().cast(nf_hook_ops_name)
+            yield nf_hook_ops
+
+
+class Netfilter(interfaces.plugins.PluginInterface):
+    """Lists Netfilter hooks."""
+
+    _required_framework_version = (2, 0, 0)
+
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Linux kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.PluginRequirement(
+                name="lsmod", plugin=lsmod.Lsmod, version=(2, 0, 0)
+            ),
+            requirements.VersionRequirement(
```

Comment:
Hmmm, should this be pulled from the abstract classes that acutally use LinuxUtilities?  Here is ok, but it's kinda split from the actual types, tagging it in there (either by comment or as a helper method) seems like it might avoid issues in the future...